File: Dynamic Programming/Java/Maximum Number of Points with Cost.java
// Problem link: https://leetcode.com/problems/maximum-number-of-points-with-cost/description
// Time: O(m * n)
// Space: O(n)
class Solution {
    public long maxPoints(int[][] points) {
        int rows = points.length, cols = points[0].length;
        long[] DP = new long[cols];
        long result = Long.MIN_VALUE;

        for(int index = 0; index < cols; index++) {
            DP[index] = points[0][index];
        }

        for(int row = 0; row < rows - 1; row++) {
            long[] left = new long[cols];
            left[0] = DP[0];

            for(int col = 1; col < cols; col++) {
                left[col] = Math.max(DP[col], left[col - 1] - 1);
            }

            DP[cols - 1] = points[row + 1][cols - 1] + left[cols - 1];

            for(int col = cols - 2; col >= 0; col--) {
                DP[col] = points[row + 1][col] + Math.max(left[col], Math.max(points[row][col], left[col + 1] - 1));
                left[col] = Math.max(left[col], Math.max(points[row][col], left[col + 1] - 1));
            }
        }

        for(int index = 0; index < cols; index++) {
            result = Math.max(result, DP[index]);
        }

        return result;
    }
}

File: Dynamic Programming/Python/Maximum Number of Points with Cost.py
# Problem link: https://leetcode.com/problems/maximum-number-of-points-with-cost/description
# Time: O(m * n)
# Space: O(n)
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        rows, cols = len(points), len(points[0])
        dp = points[0].copy()
        
        for row in range(rows - 1):
            left = [0] * cols
            left[0] = dp[0]

            for col in range(1, cols):
                left[col] = max(dp[col], left[col - 1] - 1)
            
            dp[-1] = points[row + 1][-1] + left[-1]
            
            for col in range(cols - 2, -1, -1):
                dp[col] = points[row + 1][col] + max(left[col], points[row][col], left[col + 1] - 1)
                left[col] = max(left[col], points[row][col], left[col + 1] - 1)
        
        return max(dp)
